lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -428,221 +353,294 @@ def create_image(module, ec2):
-    if img is None:
+    image = get_image_by_id(module, connection, image_id, resource, True)
+
+    if image is None:
         module.fail_json(msg="Image %s does not exist" % image_id, changed=False)
 
     try:
-        set_permissions = img.get_launch_permissions()
+        set_permissions = connection.describe_image_attribute(Attribute='launchPermission', ImageId=image_id).get('LaunchPermissions')
         if set_permissions != launch_permissions:
-            if (('user_ids' in launch_permissions and launch_permissions['user_ids']) or
-                    ('group_names' in launch_permissions and launch_permissions['group_names'])):
-                img.set_launch_permissions(**launch_permissions)
-            elif ('user_ids' in set_permissions and set_permissions['user_ids']) or ('group_names' in set_permissions and set_permissions['group_names']):
-                img.remove_launch_permissions(**set_permissions)
+            if ('user_ids' in launch_permissions or 'group_names' in launch_permissions):
+                group_names = launch_permissions.get('group_names')[0] if launch_permissions.get('group_names') else None
+                user_ids = launch_permissions.get('user_ids')[0] if launch_permissions.get('user_ids') else None
+                launch_perms_add = {'Add': [{}]}
+                if group_names:
+                    launch_perms_add['Add'][0]['Group'] = group_names
+                if user_ids:
+                    launch_perms_add['Add'][0]['UserId'] = user_ids
+                image.modify_attribute(Attribute='launchPermission', LaunchPermission=launch_perms_add)
+            elif set_permissions and set_permissions[0].get('UserId') is not None and set_permissions[0].get('Group') is not None:
+                image.modify_attribute(
+                    Attribute='launchPermission',
+                    LaunchPermission={
+                        'Remove': [{
+                            'Group': (set_permissions.get('Group') or ''),
+                            'UserId': (set_permissions.get('UserId') or '')
+                        }]
+                    })
             else:
-                module.exit_json(msg="AMI not updated", launch_permissions=set_permissions, changed=False, **get_ami_info(img))
-            module.exit_json(msg="AMI launch permissions updated", launch_permissions=launch_permissions,
-                             set_perms=set_permissions, changed=True, **get_ami_info(img))
+                module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False)
+            module.exit_json(msg="AMI launch permissions updated.", launch_permissions=launch_permissions, set_perms=set_permissions, changed=True)
         else:
-            module.exit_json(msg="AMI not updated", launch_permissions=set_permissions, changed=False, **get_ami_info(img))
+            module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False)
```

Comment:
You *MUST* return the AMI info, the RETURN document specifies that you do, and it's otherwise not backward compatible.

Reply:
reverted

Reply:
also added the following to return:
platform
enhanced_networking
image_owner_alias
image_type
kernel_id
product_codes
ramdisk_id
sriov_net_support
state_reason

if ok to keep, will add to docs section.